Objects/typevarobject.c

```diff
@@ -0,0 +1,1703 @@
+}
+
+static int
+typevar_clear(typevarobject *self)
+{
+    Py_CLEAR(self->bound);
+    Py_CLEAR(self->evaluate_bound);
+    Py_CLEAR(self->constraints);
+    Py_CLEAR(self->evaluate_constraints);
+    _PyObject_ClearManagedDict((PyObject *)self);
+    return 0;
+}
+
+static PyObject *
+typevar_repr(PyObject *self)
+{
+    typevarobject *tv = (typevarobject *)self;
+
+    if (tv->infer_variance) {
+        return PyUnicode_FromFormat("%s", tv->name);
+    }
+
+    char variance = tv->covariant ? '+' : tv->contravariant ? '-' : '~';
+    return PyUnicode_FromFormat("%c%s", variance, tv->name);
+}
+
+static PyMemberDef typevar_members[] = {
+    {"__name__", T_STRING, offsetof(typevarobject, name), READONLY},
+    {"__covariant__", T_BOOL, offsetof(typevarobject, covariant), READONLY},
+    {"__contravariant__", T_BOOL, offsetof(typevarobject, contravariant), READONLY},
+    {"__infer_variance__", T_BOOL, offsetof(typevarobject, infer_variance), READONLY},
+    {0}
+};
+
+static PyObject *
+typevar_bound(typevarobject *self, void *unused)
+{
+    if (self->bound != NULL) {
+        return Py_NewRef(self->bound);
+    }
+    if (self->evaluate_bound == NULL) {
+        Py_RETURN_NONE;
+    }
+    PyObject *bound = PyObject_CallNoArgs(self->evaluate_bound);
+    self->bound = Py_XNewRef(bound);
+    return bound;
+}
+
+static PyObject *
+typevar_constraints(typevarobject *self, void *unused)
+{
+    if (self->constraints != NULL) {
+        return Py_NewRef(self->constraints);
+    }
+    if (self->evaluate_constraints == NULL) {
+        Py_RETURN_NONE;
+    }
+    PyObject *constraints = PyObject_CallNoArgs(self->evaluate_constraints);
+    self->constraints = Py_XNewRef(constraints);
+    return constraints;
+}
+
+static PyGetSetDef typevar_getset[] = {
+    {"__bound__", (getter)typevar_bound, NULL, NULL, NULL},
+    {"__constraints__", (getter)typevar_constraints, NULL, NULL, NULL},
+    {0}
+};
+
+static typevarobject *
+typevar_alloc(const char *name, PyObject *bound, PyObject *evaluate_bound,
+              PyObject *constraints, PyObject *evaluate_constraints,
+              bool covariant, bool contravariant, bool infer_variance,
+              PyObject *module)
+{
+    PyTypeObject *tp = PyInterpreterState_Get()->cached_objects.typevar_type;
+    assert(tp != NULL);
+    typevarobject *tv = PyObject_GC_New(typevarobject, tp);
+    if (tv == NULL) {
+        return NULL;
+    }
+
+    tv->name = strdup(name);
+    if (tv->name == NULL) {
+        Py_DECREF(tv);
+        return NULL;
+    }
+
+    tv->bound = Py_XNewRef(bound);
+    tv->evaluate_bound = Py_XNewRef(evaluate_bound);
+    tv->constraints = Py_XNewRef(constraints);
+    tv->evaluate_constraints = Py_XNewRef(evaluate_constraints);
+
+    tv->covariant = covariant;
+    tv->contravariant = contravariant;
+    tv->infer_variance = infer_variance;
+    if (module != NULL) {
+        if (PyObject_SetAttrString((PyObject *)tv, "__module__", module) < 0) {
+            Py_DECREF(tv);
```

Comment:
Yes, I think we should call `PyObject_GC_Track` above.

Reply:
That makes sense. I think the same applies to the return NULL if strdup() fails, so I'll adjust that code too.

Reply:
Also reviewed the other classes in this file as a few have similar issues. I noticed another bug is that if strdup() fails (because we're out of memory), we return NULL but don't set an exception, so I'm adding `PyErr_NoMemory()` into those paths.